src/main/java/com/faforever/client/chat/KittehChatService.java

```diff
@@ -0,0 +1,539 @@
+        .map(Mode::getChar)
+        .anyMatch(character -> character == 'o');
+
+    return getOrCreateChatUser(username, channelName, isModerator);
+  }
+
+  @Subscribe
+  public void onLoginSuccessEvent(LoginSuccessEvent event) {
+    connect();
+  }
+
+  @Subscribe
+  public void onLoggedOutEvent(LoggedOutEvent event) {
+    disconnect();
+    eventBus.post(UpdateApplicationBadgeEvent.ofNewValue(0));
+  }
+
+  @Subscribe
+  public void onCurrentPlayerInfo(CurrentPlayerInfo currentPlayerInfo) {
+    if (!newbieChannelJoined && currentPlayerInfo.getCurrentPlayer().getNumberOfGames() < MAX_GAMES_FOR_NEWBIE_CHANNEL) {
+      joinChannel(NEWBIE_CHANNEL_NAME);
+    }
+    newbieChannelJoined = true;
+  }
+
+  @Subscribe
+  public void onPlayerOnline(PlayerOnlineEvent event) {
+    Player player = event.getPlayer();
+
+    synchronized (channels) {
+      List<ChatChannelUser> channelUsers = channels.values().stream()
+          .map(channel -> chatChannelUsersByChannelAndName.get(mapKey(player.getUsername(), channel.getName())))
+          .filter(Objects::nonNull)
+          .peek(chatChannelUser -> chatChannelUser.setPlayer(player))
+          .collect(Collectors.toList());
+
+      player.getChatChannelUsers().addAll(channelUsers);
+    }
+  }
+
+  @Handler
+  public void onConnect(ClientNegotiationCompleteEvent event) {
+    connectionState.set(ConnectionState.CONNECTED);
+  }
+
+  @Handler
+  private void onJoinEvent(ChannelJoinEvent event) {
+    User user = event.getActor();
+    log.debug("User joined channel: {}", user);
+    onJoinEvent(event.getChannel().getName(), getOrCreateChatUser(user, event.getChannel().getName()));
+  }
+
+  @Handler
+  public void onChatUserList(ChannelNamesUpdatedEvent event) {
+    Channel channel = event.getChannel();
+    List<ChatChannelUser> users = channel.getUsers().stream().map(user -> getOrCreateChatUser(user, channel.getName())).collect(Collectors.toList());
+    getOrCreateChannel(channel.getName()).addUsers(users);
+  }
+
+  @Handler
+  private void onPartEvent(ChannelPartEvent event) {
+    User user = event.getActor();
+    log.debug("User joined channel: {}", user);
+    onChatUserLeftChannel(event.getChannel().getName(), user.getNick());
+  }
+
+  @Handler
+  private void onMessage(ChannelMessageEvent event) {
+    User user = event.getActor();
+
+    String source;
+    Channel channel = event.getChannel();
+    source = channel.getName();
+
+    eventBus.post(new ChatMessageEvent(new ChatMessage(source, Instant.ofEpochMilli(user.getCreationTime()), user.getNick(), event.getMessage(), false)));
+  }
+
+  @Handler
+  private void onPrivateMessage(PrivateMessageEvent event) {
+    User user = event.getActor();
+    log.debug("Received private message: {}", event);
+
+    ChatChannelUser sender = getOrCreateChatUser(user.getNick(), user.getNick(), false);
+    if (sender != null
+        && sender.getPlayer().isPresent()
```

Comment:
you can skip this check if you use .map(Player::getSocialStatus) on the optional and then `== Optional.of(SocialStatus.FOE)`